### PR TITLE
Add validation step and a corresponding test for ValSem106

### LINKED ISSUE #525: Test message validation
With the new validation mechanism added in #522 in place, we should be able to better test individual validation steps. Writing these test should essentially yield a negative test vector, which we can use to test against other implementations.

### LINKED ISSUE #538: [MLS Spec change] Add group_context_extensions proposal ID
**Link to the exact changes**
https://github.com/mlswg/mls-protocol/pull/488

### PR DESCRIPTION
**Pull request by kkohbrok**
As the title says, this PR adds a validation step and a corresponding test for ValSem106.

Fixes #525.

### CODE REVIEW

File: openmls/src/group/core_group/validation.rs [hunk 1]
```diff
@@ -155,6 +157,63 @@ impl CoreGroup {
...
+            // - Do ciphersuite and version match that of the group?
+            // - Are the two listed in the `Capabilities` Extension?
+            // - If a `RequiredCapabilitiesExtension` is present in the group:
+            //   Does the key package advertise the capabilities required by that
+            //   extension?
+
+            // Check if ciphersuite and version of the group are correct.
+            if add_proposal.add_proposal().key_package().ciphersuite() != self.ciphersuite()
+                || add_proposal.add_proposal().key_package().protocol_version() != self.version()
+            {
```
Comment (franziskuskiefer) [inline openmls/src/group/core_group/validation.rs:172]: Can you log what exactly went wrong here?

File: openmls/src/group/core_group/validation.rs [hunk 2]
```diff
@@ -155,6 +157,63 @@ impl CoreGroup {
...
+                    // Mismatches between Extensions and ExtensionTypes should be
+                    // caught when constructing KeyPackages.
+                    ProposalValidationError::LibraryError(LibraryError::custom(
+                        "ExtensionType didn't match extension content.",
+                    ))
+                })?;
+            if !capabilities.ciphersuites().contains(&self.ciphersuite())
+                || !capabilities.versions().contains(&self.version())
+            {
+                return Err(ProposalValidationError::InsufficientCapabilities);
```
Comment (franziskuskiefer) [inline openmls/src/group/core_group/validation.rs:194]: Sam here. I know that's not done in most places yet, but it really should. The error message won't tell us much later.

File: openmls/src/group/core_group/validation.rs [hunk 3]
```diff
@@ -155,6 +157,63 @@ impl CoreGroup {
...
+                    .map_err(|_| {
+                        // Mismatches between Extensions and ExtensionTypes should be
+                        // caught when constructing KeyPackages.
+                        ProposalValidationError::LibraryError(LibraryError::custom(
+                            "ExtensionType didn't match extension content.",
+                        ))
+                    })?;
+                // Check if all required capabilities are supported.
+                if !capabilities.supports_required_capabilities(required_capabilities) {
+                    return Err(ProposalValidationError::InsufficientCapabilities);
```
Comment (franziskuskiefer) [inline openmls/src/group/core_group/validation.rs:214]: And here.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks, generally looks good but let's add some logging.

**Review by franziskuskiefer — APPROVED** [review]

**Review by raphaelrobert — APPROVED** [review]
Thanks for the PR!

**Pull request merged**